zetaclient/chains/evm/observer/v2_inbound.go

```diff
@@ -0,0 +1,175 @@
+package observer
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"sort"
+
+	sdkmath "cosmossdk.io/math"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/zeta-chain/protocol-contracts/v2/pkg/gatewayevm.sol"
+
+	"github.com/zeta-chain/zetacore/pkg/coin"
+	"github.com/zeta-chain/zetacore/pkg/constant"
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/evm"
+	"github.com/zeta-chain/zetacore/zetaclient/compliance"
+	"github.com/zeta-chain/zetacore/zetaclient/config"
+	"github.com/zeta-chain/zetacore/zetaclient/metrics"
+	"github.com/zeta-chain/zetacore/zetaclient/zetacore"
+)
+
+// ObserveGateway queries the gateway contract for deposit/call events
+// returns the last block successfully scanned
+func (ob *Observer) ObserveGateway(ctx context.Context, startBlock, toBlock uint64) uint64 {
```

Comment:
I think this method should return `(uint64, error)` and use **sentinel** errors inside. 

Callers of `ob.ObserverGateway()` should handle those errors. Ideally we also want to have a return of some kind of `outcome` enum. i.e. `noop`, `processed`, `processedPartially`, `...` i.e. `(uint64, outcome, error)`

Reply:
I put the error handling at a higher level as cleaner: [63c4698](https://github.com/zeta-chain/node/pull/2616/commits/63c4698c7cad0beafb23035692782cf52c590fcf)

For the other changes containing `outcome`, I agree but I think it can be part of a separate PR for general refactoring of the observation logic. I put a TODO for https://github.com/zeta-chain/node/issues/2493 in the comments